src/blog/memory-safe-languages/myths-zero-overhead-memory-safety.vmd

```diff
@@ -0,0 +1,308 @@
+
+In this little journey, we've come across six kinds of overhead:
+
+ * Reference counting
+ * Tracing
+ * Bounds checking
+ * Cloning
+ * Hashing
+ * Generation checks
+
+
+These costs are not all equal, though. The first two are much more expensive:
+
+ * For reference counting, when we make a new reference to an object, we need to reach all the way into that object to increment its reference count number. This is often a cache miss.
+ * For tracing garbage collection, we need to crawl through all "object roots" and following their member references to other objects, and follow _their_ member references to other objects, and so on to find all the live objects to keep alive past the next collection. This involves a lot of cache misses.
+
+
+A *cache miss* is when the CPU requests memory from a 64B chunk ("cache line") that's already in the CPU cache, and it has to wait idle for hundreds of cycles waiting for the data to come all the way from RAM.
+
+
+The latter four don't have as many cache misses, they're generally much faster than reference counting and tracing garbage collection.
+
+
+Though, they still have their costs:
+
+ * Bounds checking is likely on the lower side of [1-15%](https://shnatsel.medium.com/how-to-avoid-bounds-checks-in-rust-without-unsafe-f65e618b4c1e). My basic cellular automata program showed 18%. An [interpreter](https://ceronman.com/2021/07/22/my-experience-crafting-an-interpreter-with-rust/), a rather extreme case, showed 74% overhead.
+ * Cloning can be expensive if we're cloning a string or a vector, which involves some heap allocation and its cache misses.
+ * Hash maps could be expensive depending on the hashing algorithms, such as ones that use modulus.
+ * Generational references access the object's field in parallel with its generation which is likely on the same cache line, but it can be expensive in tight loops unless one uses [linear style](https://verdagon.dev/blog/linear-types-borrowing).
+
+
+To really get good performance, one has to *pick their poison* and decide what overhead they're willing to suffer.
+
+
+Though if we're lucky, a particular situation has something special that makes some approaches less costly than others, which is always nice.
+
+
+<ignore>
+#factorio: The game [Factorio](https://www.factorio.com/) is a marvel of modern optimization. Someone once asked,
+
+> How can you prevent "random access" to memory when those inserters and assemblers are stored in different lists and presumably in different indexes in those lists?
+
+One of the Factorio developers [replied](https://www.reddit.com/r/factorio/comments/13bsf3s/comment/jjcra0t/),
+
+> You can't; it's the nature of complex systems; they end up interacting with other systems. In the end you just have to pay the cost to access that memory if you want the inserters to do what inserters do.
+</ignore>
+
+
+<slice />
+
+# What does it all mean?
+
+There's no way to get to zero-overhead memory safety.
+
+However, *we can get close.*
```

Comment:
My conclusion is that if your goal is performance, then you have to provide a lot of options to the programmer so they can select for their use case. Even GCing can be fast in some cases for instance.

This is also the idea behind Cone btw.